<!doctype html>
<html class="no-js">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark">
        <script async src="https://www.googletagmanager.com/gtag/js?id=G-ZKD1FQDEYH"></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            
            gtag('config', 'G-ZKD1FQDEYH');
            
        </script>
    <link rel="index" title="Index" href="../../../../../genindex.html" /><link rel="search" title="Search" href="../../../../../search.html" />

    <meta name="generator" content="sphinx-4.4.0, furo 2022.02.14.1"/>
        <title>graphein.protein.features.sequence.propy - Graphein 1.4.0</title>
      <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/styles/furo.css?digest=a8f031f701c6af50c9919b48f173a5785af06fac" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/styles/furo-extensions.css?digest=25ceb02ed1c46dc30f2321ff83e92799f69dfdb9" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  body[data-theme="dark"] {
    --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
  }
  @media (prefers-color-scheme: dark) {
    body:not([data-theme="light"]) {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
  }
</style></head>
  <body>
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../../../index.html"><div class="brand">Graphein 1.4.0</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand centered" href="../../../../../index.html">
  
  <div class="sidebar-logo-container">
    <img class="sidebar-logo" src="../../../../../_static/graphein.png" alt="Logo"/>
  </div>
  
  <span class="sidebar-brand-text">Graphein 1.4.0</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../../../search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../getting_started/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../getting_started/introduction.html">Graphein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../getting_started/usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../license.html">License</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../protein_notebooks.html">Protein</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/residue_graphs.html">Graphein - Residue Graph Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/atom_graph_tutorial.html">Graphein - Atom Graph Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/alphafold_protein_graph_tutorial.html">Constructing Graphs from AlphaFold Protein Structure Database</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/higher_order_graphs.html">Secondary Structure Graphs &amp; Chain Graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/subgraphing_tutorial.html">Constructing Subgraphs in Graphein</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/protein_mesh_tutorial.html">Creating Protein Meshes in Graphein &amp; 3D Visualisation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/protein_graph_analytics.html">Protein Graph Analytics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/interactive_plotly_example.html">Interactive Visualisation with Plotly</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../molecule_notebooks.html">Molecules</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/molecule_tutorial.html">Creating Molecular Graphs in Graphein</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../rna_notebooks.html">RNA Graphs</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/rna_graph_tutorial.html">Graphein RNA Graph Construction Tutorial</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../ppi_notebooks.html">PPI Networks</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/ppi_tutorial.html">PPI Networks and Structural Interactomics</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../grn_notebooks.html">Gene Regulatory Networks</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/grn_tutorial.html">Gene Regulatory Network Tutorial</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Machine Learning</span></p>
<ul>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../../../datasets.html">Datasets</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle child pages in navigation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/dataloader_tutorial.html">Graphein Protein Structure Dataloaders</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../datasets/pscdb.html">PSCDB - Protein Structural Change Database</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../notebooks/pscdb_baselines.html">PSCDB - Baselines</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../../../datasets/ppisp.html">PPISP - Protein Protein Interaction Structure Prediction</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.protein.html">graphein.protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.molecule.html">graphein.molecule</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.rna.html">graphein.rna</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.ppi.html">graphein.ppi</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.grn.html">graphein.grn</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.ml.html">graphein.ml</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../modules/graphein.utils.html">graphein.utils</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Contributing</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../contributing/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../contributing/code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../contributing/contributors.html">Contributors</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <div class="content-icon-container"><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for graphein.protein.features.sequence.propy</h1><div class="highlight"><pre>
<span></span><span class="c1"># %%</span>
<span class="c1"># Graphein</span>
<span class="c1"># Author: Arian Jamasb &lt;arian@jamasb.io&gt;, Eric Ma</span>
<span class="c1"># License: MIT</span>
<span class="c1"># Project Website: https://github.com/a-r-j/graphein</span>
<span class="c1"># Code Repository: https://github.com/a-r-j/graphein</span>

<span class="c1"># TODO: This the docstrings inside this module's set of functions</span>
<span class="c1"># can do without the `type`s.</span>
<span class="c1"># This is because the types are already annotated on the function.</span>

<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span>

<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>

<span class="kn">from</span> <span class="nn">graphein.protein.features.sequence.utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">aggregate_feature_over_chains</span><span class="p">,</span>
    <span class="n">compute_feature_over_chains</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">graphein.protein.features.utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">aggregate_graph_feature_over_chains</span><span class="p">,</span>
    <span class="n">convert_graph_dict_feat_to_series</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="compute_propy_feature"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.compute_propy_feature">[docs]</a><span class="k">def</span> <span class="nf">compute_propy_feature</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span>
    <span class="n">func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span>
    <span class="n">feature_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Computes Propy Descriptors over chains in a Protein Graph</span>

<span class="sd">    :param G: Protein Graph</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param func: ProPy wrapper function to compute</span>
<span class="sd">    :type func: Callable</span>
<span class="sd">    :param feature_name: Name of feature to index it in the nx.Graph object</span>
<span class="sd">    :type feature_name: str</span>
<span class="sd">    :param aggregation_type: Type of aggregation to use when aggregating a feature over multiple chains. One of:</span>
<span class="sd">        ["mean", "man", "sum"]. Defaults to None.</span>
<span class="sd">    :type aggregation_type: List[str], optional</span>
<span class="sd">    :return G: Returns protein Graph with features added. Features are accessible with G.graph[{feature_name}_{chain | aggegation_type}]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">)</span>

    <span class="c1"># Convert to Series</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">convert_graph_dict_feat_to_series</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">)</span>

    <span class="c1"># Aggregate features</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">graph</span><span class="p">[</span><span class="s2">"chain_ids"</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">aggregation_type</span><span class="p">:</span>
            <span class="n">G</span> <span class="o">=</span> <span class="n">aggregate_graph_feature_over_chains</span><span class="p">(</span>
                <span class="n">G</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">log</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">"Aggregation not carried out on single-chain graph"</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">G</span></div>


<div class="viewcode-block" id="amino_acid_composition"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.amino_acid_composition">[docs]</a><span class="k">def</span> <span class="nf">amino_acid_composition</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the composition of Amino acids for a given protein sequence.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use</span>
<span class="sd">    :type aggregation_type: Optional[List[str]]</span>
<span class="sd">    :return: Protein Graph with amino_acid_composition feature added. G.graph["amino_acid_composition_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.AAComposition</span> <span class="kn">import</span> <span class="n">CalculateAAComposition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateAAComposition</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"amino_acid_composition"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="dipeptide_composition"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.dipeptide_composition">[docs]</a><span class="k">def</span> <span class="nf">dipeptide_composition</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the composition of dipeptidefor a given protein sequence. Contains composition of 400 dipeptides</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with dipeptide_composition feature added. G.graph["dipeptide_composition_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.AAComposition</span> <span class="kn">import</span> <span class="n">CalculateDipeptideComposition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDipeptideComposition</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"dipeptide_composition"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="aa_dipeptide_composition"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.aa_dipeptide_composition">[docs]</a><span class="k">def</span> <span class="nf">aa_dipeptide_composition</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the composition of AADs, dipeptide and 3-mers for a given protein sequence. Contains all composition</span>
<span class="sd">        values of AADs, dipeptide and 3-mers (8420).</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with aa_dipeptide_composition feature added.</span>
<span class="sd">        G.graph["aa_dipeptide_composition_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.AAComposition</span> <span class="kn">import</span> <span class="n">CalculateAADipeptideComposition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateAADipeptideComposition</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"aa_dipeptide_composition"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="aa_spectrum"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.aa_spectrum">[docs]</a><span class="k">def</span> <span class="nf">aa_spectrum</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the spectrum descriptors of 3-mers for a given protein. Contains the composition values of 8000 3-mers</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with aa_spectrum feature added. G.graph["aa_spectrum_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.AAComposition</span> <span class="kn">import</span> <span class="n">GetSpectrumDict</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">GetSpectrumDict</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"aa_spectrum"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<span class="c1">######## CTD Funcs #######</span>


<div class="viewcode-block" id="all_composition_descriptors"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.all_composition_descriptors">[docs]</a><span class="k">def</span> <span class="nf">all_composition_descriptors</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate all composition descriptors based on seven different properties of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_descriptors feature added.</span>
<span class="sd">        G.graph["composition_descriptors_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateC</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateC</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_descriptors"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="all_ctd_descriptors"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.all_ctd_descriptors">[docs]</a><span class="k">def</span> <span class="nf">all_ctd_descriptors</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate all CTD descriptors based seven different properties of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with ctd_descriptors feature added.</span>
<span class="sd">        G.graph["ctd_descriptors_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCTD</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCTD</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"ctd_descriptors"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_descriptor"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_descriptor">[docs]</a><span class="k">def</span> <span class="nf">composition_descriptor</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span>
    <span class="n">AAProperty</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">AAPName</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute composition descriptors.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param AAProperty: contains classification of amino acids such as _Polarizability.</span>
<span class="sd">    :type AAProperty: Dict[Any, Any]</span>
<span class="sd">    :param AAPName: used for indicating a AAP name.</span>
<span class="sd">    :type AAPName: str</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_{AAPName} feature added.</span>
<span class="sd">        G.graph["composition_{AAPName}_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateComposition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span>
        <span class="n">CalculateComposition</span><span class="p">,</span> <span class="n">AAProperty</span><span class="o">=</span><span class="n">AAProperty</span><span class="p">,</span> <span class="n">AAPName</span><span class="o">=</span><span class="n">AAPName</span>
    <span class="p">)</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">"composition_</span><span class="si">{</span><span class="n">AAPName</span><span class="si">}</span><span class="s2">"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_charge"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_charge">[docs]</a><span class="k">def</span> <span class="nf">composition_charge</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on Charge of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_charge feature added.</span>
<span class="sd">        G.graph["composition_charge_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionCharge</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionCharge</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_charge"</span>

    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"composition_charge"</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">composition_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on Hydrophobicity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_hydrophobicity feature added.</span>
<span class="sd">        G.graph["composition_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionHydrophobicity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_normalized_vdwv"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_normalized_vdwv">[docs]</a><span class="k">def</span> <span class="nf">composition_normalized_vdwv</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on NormalizedVDWV of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_normalized_vdwv feature added.</span>
<span class="sd">        G.graph["composition_normalized_vdwv_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionNormalizedVDWV</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionNormalizedVDWV</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_normalised_vdwv"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_polarity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_polarity">[docs]</a><span class="k">def</span> <span class="nf">composition_polarity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on Polarity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_polarity feature added.</span>
<span class="sd">        G.graph["composition_polarity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionPolarity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionPolarity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_polarity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_polarizability">[docs]</a><span class="k">def</span> <span class="nf">composition_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on Polarizability of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_polarizability feature added.</span>
<span class="sd">        G.graph["composition_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionPolarizability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_secondary_str"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_secondary_str">[docs]</a><span class="k">def</span> <span class="nf">composition_secondary_str</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on SecondaryStr of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_secondary_str feature added.</span>
<span class="sd">        G.graph["composition_secondary_str_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionSecondaryStr</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionSecondaryStr</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_secondary_str"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="composition_solvent_accessibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.composition_solvent_accessibility">[docs]</a><span class="k">def</span> <span class="nf">composition_solvent_accessibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate composition descriptors based on SolventAccessibility of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with composition_solvent_accessibility feature added.</span>
<span class="sd">        G.graph["composition_solvent_accessibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateCompositionSolventAccessibility</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateCompositionSolventAccessibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"composition_solvent_accessibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="all_distribution_descriptors"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.all_distribution_descriptors">[docs]</a><span class="k">def</span> <span class="nf">all_distribution_descriptors</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate all distribution descriptors based on seven different properties of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_descriptors feature added.</span>
<span class="sd">        G.graph["distribution_descriptors_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateC</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateC</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_descriptors"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_descriptor"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_descriptor">[docs]</a><span class="k">def</span> <span class="nf">distribution_descriptor</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span>
    <span class="n">AAProperty</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">AAPName</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute distribution descriptors.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param AAProperty: contains classifciation of amino acids such as _Polarizability.</span>
<span class="sd">    :type AAProperty: Dict[Any, Any]</span>
<span class="sd">    :param AAPName: used for indicating a AAP name.</span>
<span class="sd">    :type AAPName: str</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_{AAPName} feature added.</span>
<span class="sd">        G.graph["distribution_{AAPName}_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistribution</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span>
        <span class="n">CalculateDistribution</span><span class="p">,</span> <span class="n">AAProperty</span><span class="o">=</span><span class="n">AAProperty</span><span class="p">,</span> <span class="n">AAPName</span><span class="o">=</span><span class="n">AAPName</span>
    <span class="p">)</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">"distribution_</span><span class="si">{</span><span class="n">AAPName</span><span class="si">}</span><span class="s2">"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_charge"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_charge">[docs]</a><span class="k">def</span> <span class="nf">distribution_charge</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on Charge of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_charge feature added.</span>
<span class="sd">        G.graph["distribution_charge_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionCharge</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionCharge</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_charge"</span>

    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"distribution_charge"</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">distribution_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on Hydrophobicity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_hydrophobicity feature added.</span>
<span class="sd">        G.graph["distribution_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionHydrophobicity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_normalized_vdwv"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_normalized_vdwv">[docs]</a><span class="k">def</span> <span class="nf">distribution_normalized_vdwv</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on NormalizedVDWV of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_normalized_vdwv feature added.</span>
<span class="sd">        G.graph["distribution_normalized_vdwv_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionNormalizedVDWV</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionNormalizedVDWV</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_normalised_vdwv"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_polarity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_polarity">[docs]</a><span class="k">def</span> <span class="nf">distribution_polarity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on Polarity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_polarity feature added.</span>
<span class="sd">        G.graph["distribution_polarity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionPolarity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionPolarity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_polarity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_polarizability">[docs]</a><span class="k">def</span> <span class="nf">distribution_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on Polarizability of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_polarizability feature added.</span>
<span class="sd">        G.graph["distribution_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionPolarizability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_secondary_str"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_secondary_str">[docs]</a><span class="k">def</span> <span class="nf">distribution_secondary_str</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on SecondaryStr of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_secondary_str feature added.</span>
<span class="sd">        G.graph["distribution_secondary_str_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionSecondaryStr</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionSecondaryStr</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_secondary_str"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="distribution_solvent_accessibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.distribution_solvent_accessibility">[docs]</a><span class="k">def</span> <span class="nf">distribution_solvent_accessibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate distribution descriptors based on SolventAccessibility of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with distribution_solvent_accessibility feature added.</span>
<span class="sd">        G.graph["distribution_solvent_accessibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateDistributionSolventAccessibility</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateDistributionSolventAccessibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"distribution_solvent_accessibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="all_transition_descriptors"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.all_transition_descriptors">[docs]</a><span class="k">def</span> <span class="nf">all_transition_descriptors</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate all transition descriptors based on seven different properties of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_descriptors feature added.</span>
<span class="sd">        G.graph["transition_descriptors_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateC</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateC</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_descriptors"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_descriptor"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_descriptor">[docs]</a><span class="k">def</span> <span class="nf">transition_descriptor</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span>
    <span class="n">AAProperty</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">AAPName</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute transition descriptors.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param AAProperty: contains classifciation of amino acids such as _Polarizability.</span>
<span class="sd">    :type AAProperty: Dict[Any, Any]</span>
<span class="sd">    :param AAPName: used for indicating a AAP name.</span>
<span class="sd">    :type AAPName: str</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_{AAPName} feature added.</span>
<span class="sd">        G.graph["transition_{AAPName}_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">CalculateTransition</span><span class="p">,</span> <span class="n">AAProperty</span><span class="o">=</span><span class="n">AAProperty</span><span class="p">,</span> <span class="n">AAPName</span><span class="o">=</span><span class="n">AAPName</span><span class="p">)</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">"transition_</span><span class="si">{</span><span class="n">AAPName</span><span class="si">}</span><span class="s2">"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_charge"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_charge">[docs]</a><span class="k">def</span> <span class="nf">transition_charge</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on Charge of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_charge feature added.</span>
<span class="sd">        G.graph["transition_charge_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionCharge</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionCharge</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_charge"</span>

    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"transition_charge"</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">transition_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on Hydrophobicity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_hydrophobicity feature added.</span>
<span class="sd">        G.graph["transition_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionHydrophobicity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_normalized_vdwv"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_normalized_vdwv">[docs]</a><span class="k">def</span> <span class="nf">transition_normalized_vdwv</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on NormalizedVDWV of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_normalized_vdwv feature added.</span>
<span class="sd">        G.graph["transition_normalized_vdwv_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionNormalizedVDWV</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionNormalizedVDWV</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_normalised_vdwv"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_polarity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_polarity">[docs]</a><span class="k">def</span> <span class="nf">transition_polarity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on Polarity of AADs.</span>

<span class="sd">    :param G: Protein Graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_polarity feature added.</span>
<span class="sd">        G.graph["transition_polarity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionPolarity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionPolarity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_polarity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_polarizability">[docs]</a><span class="k">def</span> <span class="nf">transition_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on Polarizability of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_polarizability feature added.</span>
<span class="sd">        G.graph["transition_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionPolarizability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_secondary_str"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_secondary_str">[docs]</a><span class="k">def</span> <span class="nf">transition_secondary_str</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on SecondaryStr of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_secondary_str feature added.</span>
<span class="sd">        G.graph["transition_secondary_str_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionSecondaryStr</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionSecondaryStr</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_secondary_str"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="transition_solvent_accessibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.transition_solvent_accessibility">[docs]</a><span class="k">def</span> <span class="nf">transition_solvent_accessibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate transition descriptors based on SolventAccessibility of AADs.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with transition_solvent_accessibility feature added.</span>
<span class="sd">        G.graph["transition_solvent_accessibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.CTD</span> <span class="kn">import</span> <span class="n">CalculateTransitionSolventAccessibility</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateTransitionSolventAccessibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"transition_solvent_accessibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<span class="c1">####### Autocorrelation descriptors ########</span>


<div class="viewcode-block" id="autocorrelation_total"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_total">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_total</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute all autocorrelation descriptors based on 8 properties of AADs. result contains 30*8*3=720 normalized Moreau</span>
<span class="sd">        Broto, Moran, and Geary</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_total feature added.</span>
<span class="sd">        G.graph["autocorrelation_total_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateAutoTotal</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateAutoTotal</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_total"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_all"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_all">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_all</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute Geary autocorrelation descriptors based on 8 properties of AADs. Result contains 30*8=240 Geary</span>
<span class="sd">        autocorrelation descriptors based on the given properties(i.e., _AAPropert).</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_all feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_all_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoTotal</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoTotal</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_all"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_av_flexibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_av_flexibility">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_av_flexibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on AvFlexibility. contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on AvFlexibility.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_av_flexibility feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_av_flexibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoAvFlexibility</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoAvFlexibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_av_flexibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_free_energy"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_free_energy">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_free_energy</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on FreeEnergy. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on FreeEnergy.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_free_energy feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_av_free_energy_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoFreeEnergy</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoFreeEnergy</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_free_energy"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on hydrophobicity. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on hydrophobicity.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_hydrophobicity feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoHydrophobicity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_mutability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_mutability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_mutability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on Mutability. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on mutability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_mutability feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_mutability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoMutability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoMutability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_mutability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_polarizability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on polarizability. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on polarizability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_polarizability feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoPolarizability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_residue_asa"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_residue_asa">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_residue_asa</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on ResidueASA. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on ResidueASA.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_residue_asa feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_residue_asa_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoResidueASA</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoResidueASA</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_residue_asa"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_residue_vol"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_residue_vol">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_residue_vol</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on ResidueVol. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on ResidueVol.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_residue_vol feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_residue_vol_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoResidueVol</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoResidueVol</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_residue_vol"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_geary_steric"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_geary_steric">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_geary_steric</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Geary Autocorrelation descriptors based on Steric. result contains 30 Geary Autocorrelation</span>
<span class="sd">        descriptors based on Steric</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_geary_steric feature added.</span>
<span class="sd">        G.graph["autocorrelation_geary_steric_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateGearyAutoSteric</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateGearyAutoSteric</span><span class="p">()</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_geary_steric"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_all"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_all">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_all</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute Moran autocorrelation descriptors based on 8 properties of AADs. Result contains 30*8=240 Moran</span>
<span class="sd">        autocorrelation descriptors based on the given properties(i.e., _AAPropert).</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_all feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_all_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoTotal</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoTotal</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_all"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_av_flexibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_av_flexibility">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_av_flexibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on AvFlexibility. contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on AvFlexibility.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_av_flexibility feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_av_flexibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoAvFlexibility</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoAvFlexibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_av_flexibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_free_energy"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_free_energy">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_free_energy</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on FreeEnergy. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on FreeEnergy.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_free_energy feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_av_free_energy_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoFreeEnergy</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoFreeEnergy</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_free_energy"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on hydrophobicity. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on hydrophobicity.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_hydrophobicity feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoHydrophobicity</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_mutability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_mutability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_mutability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on Mutability. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on mutability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_mutability feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_mutability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoMutability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoMutability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_mutability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_polarizability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on polarizability. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on polarizability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_polarizability feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoPolarizability</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_residue_asa"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_residue_asa">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_residue_asa</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on ResidueASA. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on ResidueASA.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_residue_asa feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_residue_asa_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoResidueASA</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoResidueASA</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_residue_asa"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_residue_vol"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_residue_vol">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_residue_vol</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on ResidueVol. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on ResidueVol.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_residue_vol feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_residue_vol_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoResidueVol</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoResidueVol</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_residue_vol"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_moran_steric"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_moran_steric">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_moran_steric</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the Moran Autocorrelation descriptors based on Steric. result contains 30 Moran Autocorrelation</span>
<span class="sd">        descriptors based on Steric</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_moran_steric feature added.</span>
<span class="sd">        G.graph["autocorrelation_moran_steric_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateMoranAutoSteric</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateMoranAutoSteric</span><span class="p">()</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_moran_steric"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_all"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_all">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_all</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute NormalizedMoreauBroto autocorrelation descriptors based on 8 properties of AADs. Result contains 30*8=240</span>
<span class="sd">        NormalizedMoreauBroto autocorrelation descriptors based on the given properties(i.e., _AAPropert).</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_all feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_all_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateNormalizedMoreauBrotoAutoTotal</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoTotal</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_all"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_av_flexibility"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_av_flexibility">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_av_flexibility</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on AvFlexibility. contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on AvFlexibility.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_av_flexibility feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_av_flexibility_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoAvFlexibility</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoAvFlexibility</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_av_flexibility"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_free_energy"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_free_energy">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_free_energy</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on FreeEnergy. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on FreeEnergy.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_free_energy feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_av_free_energy_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoFreeEnergy</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoFreeEnergy</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_free_energy"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_hydrophobicity"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_hydrophobicity">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_hydrophobicity</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on hydrophobicity. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on hydrophobicity.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_hydrophobicity feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_hydrophobicity_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoHydrophobicity</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoHydrophobicity</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_hydrophobicity"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_mutability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_mutability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_mutability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on Mutability. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on mutability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_mutability feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_mutability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoMutability</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoMutability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_mutability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_polarizability"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_polarizability">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_polarizability</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on polarizability. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on polarizability.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_polarizability feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_polarizability_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoPolarizability</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoPolarizability</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_polarizability"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_residue_asa"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_residue_asa">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_residue_asa</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on ResidueASA. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on ResidueASA.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_residue_asa feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_residue_asa_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoResidueASA</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoResidueASA</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_residue_asa"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_residue_vol"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_residue_vol">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_residue_vol</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on ResidueVol. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on ResidueVol.</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_residue_vol feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_residue_vol_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">CalculateNormalizedMoreauBrotoAutoResidueVol</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoResidueVol</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_residue_vol"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="autocorrelation_normalized_moreau_broto_steric"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.autocorrelation_normalized_moreau_broto_steric">[docs]</a><span class="k">def</span> <span class="nf">autocorrelation_normalized_moreau_broto_steric</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">aggregation_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Calculate the NormalizedMoreauBroto Autocorrelation descriptors based on Steric. result contains 30</span>
<span class="sd">        NormalizedMoreauBroto Autocorrelation descriptors based on Steric</span>

<span class="sd">    :param G: Protein graph to featurise</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param aggregation_type: Aggregation types to use over chains</span>
<span class="sd">    :type aggregation_type: List[Optional[str]]</span>
<span class="sd">    :return: Protein Graph with autocorrelation_normalized_moreau_broto_steric feature added.</span>
<span class="sd">        G.graph["autocorrelation_normalized_moreau_broto_steric_{chain | aggregation_type}"]</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.Autocorrelation</span> <span class="kn">import</span> <span class="n">CalculateNormalizedMoreauBrotoAutoSteric</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">CalculateNormalizedMoreauBrotoAutoSteric</span><span class="p">()</span>
    <span class="n">feature_name</span> <span class="o">=</span> <span class="s2">"autocorrelation_normalized_moreau_broto_steric"</span>

    <span class="k">return</span> <span class="n">compute_propy_feature</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
        <span class="n">feature_name</span><span class="o">=</span><span class="n">feature_name</span><span class="p">,</span>
        <span class="n">aggregation_type</span><span class="o">=</span><span class="n">aggregation_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<span class="c1">####### Quasi Sequence Order Descriptors ######</span>


<span class="k">def</span> <span class="nf">quasi_sequence_order_aa_composition</span><span class="p">(</span><span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">propy.QuasiSequenceOrder</span> <span class="kn">import</span> <span class="n">GetAAComposition</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">GetAAComposition</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"quasi_seq_order_aa_composition"</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">G</span>


<div class="viewcode-block" id="quasi_sequence_order"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.quasi_sequence_order">[docs]</a><span class="k">def</span> <span class="nf">quasi_sequence_order</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">maxlag</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">weight</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute quasi-sequence-order descriptors for a given protein.</span>

<span class="sd">    Kuo-Chen Chou. Prediction of Protein Subcellar Locations by Incorporating Quasi-Sequence-Order Effect.</span>
<span class="sd">        Biochemical and Biophysical Research Communications 2000, 278, 477-483.</span>

<span class="sd">    :param maxlag: (int, optional (default: 30))  the maximum lag and the length of the protein should be larger than</span>
<span class="sd">        maxlag</span>
<span class="sd">    :type maxlag: int</span>
<span class="sd">    :param weight: (float, optional (default: 0.1))  a weight factor. Please see reference 1 for its choice.</span>
<span class="sd">    :type weight: float</span>
<span class="sd">    :returns:</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.QuasiSequenceOrder</span> <span class="kn">import</span> <span class="n">GetQuasiSequenceOrder</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">GetQuasiSequenceOrder</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="n">weight</span><span class="p">,</span> <span class="n">maxlag</span><span class="o">=</span><span class="n">maxlag</span><span class="p">)</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"quasi_sequence_order"</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">G</span></div>


<div class="viewcode-block" id="sequence_order_coupling_number_total"><a class="viewcode-back" href="../../../../../modules/graphein.protein.html#graphein.protein.features.sequence.propy.sequence_order_coupling_number_total">[docs]</a><span class="k">def</span> <span class="nf">sequence_order_coupling_number_total</span><span class="p">(</span>
    <span class="n">G</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">,</span> <span class="n">maxlag</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
    <span class="sd">"""</span>
<span class="sd">    Compute the sequence order coupling numbers from 1 to maxlag for a given protein sequence.</span>

<span class="sd">    :param G: Protein Graph</span>
<span class="sd">    :type G: nx.Graph</span>
<span class="sd">    :param maxlag (int, optional (default: 30))  the maximum lag and the length of the protein should be larger</span>
<span class="sd">    :type maxlag: int</span>
<span class="sd">    :returns:</span>
<span class="sd">    :rtype: nx.Graph</span>
<span class="sd">    """</span>
    <span class="kn">from</span> <span class="nn">propy.QuasiSequenceOrder</span> <span class="kn">import</span> <span class="n">GetSequenceOrderCouplingNumberTotal</span>

    <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">GetSequenceOrderCouplingNumberTotal</span><span class="p">,</span> <span class="n">maxlag</span><span class="o">=</span><span class="n">maxlag</span><span class="p">)</span>
    <span class="n">G</span> <span class="o">=</span> <span class="n">compute_feature_over_chains</span><span class="p">(</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">func</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">"sequence_order_coupling_number_total"</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">G</span></div>


<span class="c1"># Todo feature aggregation</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">"__main__"</span><span class="p">:</span>
    <span class="k">pass</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, Arian Jamasb
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../../../" id="documentation_options" src="../../../../../_static/documentation_options.js"></script>
    <script src="../../../../../_static/jquery.js"></script>
    <script src="../../../../../_static/underscore.js"></script>
    <script src="../../../../../_static/doctools.js"></script>
    <script src="../../../../../_static/scripts/furo.js"></script>
    <script src="../../../../../_static/clipboard.min.js"></script>
    <script src="../../../../../_static/copybutton.js"></script>
    <script src="../../../../../_static/tabs.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>